Draft for discussion

Piinks opened the pull request:
Will add context comments before our sync up

Piinks committed:
Commit to switch

Piinks committed:
Stopping

Piinks commented on lib/components/CustomNavigationBar.dart:
```diff
@@ -32,98 +32,92 @@ class _CustomNavigationBarState extends State<CustomNavigationBar> {
-                    ),
-                  ),
-                ),
+        child: Stack(
```
Extra Containers and Padding weren't needed here

Piinks commented on lib/components/CustomSwitch.dart:
```diff
@@ -119,14 +120,14 @@ class Track extends StatelessWidget {
-        decoration: BoxDecoration(
-            borderRadius: $styles.borderStyle.trackBorderRadius,
-            color: widget.value ? activeColor : inactiveColor,
-            border: Border.all(
-                color: $styles.borderStyle.stroke,
-                width: $styles.borderStyle.width)));
```
Noticed a lot of code duplication of this across the app, consolidated in $styles to reduce.

Piinks commented on lib/components/Header.dart:
```diff
@@ -17,38 +17,34 @@ class CustomSliverHeaderDelegate extends SliverPersistentHeaderDelegate {
-                        child: const Text("In season today")),
-                  ]),
-            ),
+          padding: EdgeInsets.only(
```
Container has Padding.
So any `Container(child: Padding...` can be reduced.

Piinks commented on lib/components/Header.dart:
```diff
@@ -17,38 +17,34 @@ class CustomSliverHeaderDelegate extends SliverPersistentHeaderDelegate {
           color: $styles.colors.background,
-          child: Padding(
-            padding: EdgeInsets.symmetric(horizontal: $styles.padding.l),
-            child: Padding(
```
Padding inside of Padding wasn't necessary, can just be one.

Piinks commented on lib/components/NumberedCircle.dart:
```diff
@@ -18,26 +18,24 @@ class NumberedCircle extends StatelessWidget {
-          : null,
+      child: Center(
+        child: Text(
+          text,
+          style: text.length > 2 ? $styles.text.heading3 : $styles.text.heading2,
```
Just simple reductions here based on text not being nullable and having a fallback value.

khanhnwin replied:
I've been writing too many null checks that I didn't need to write -.-

Piinks commented on lib/components/Search.dart:
```diff
@@ -27,35 +36,20 @@ class Search extends StatelessWidget {
-        prefixIcon: SizedBox(
-          width: 20,
-          height: 20,
+        prefixIcon: SizedBox.square(
```
SizedBox.square helped reduce a bunch across the app instead of specifying the same height and width

Piinks commented on lib/components/Search.dart:
```diff
@@ -27,35 +36,20 @@ class Search extends StatelessWidget {
-              width: $styles.borderStyle.width,
-            ),
-            borderRadius: $styles.borderStyle.circularBorderRadius),
+        // What did not work? We should file a bug.
```
What is not working here? Can we file a bug to investigate?

khanhnwin replied:
@leighajarett I don't believe I touched the Search code! You probably have better insight than me!

leighajarett replied:
I think I was confused that "border" didn't set the border for all the different states but seems like it's intended - I can still file a bug to track it

leighajarett replied:
Added here: https://github.com/flutter/flutter/issues/140331

Piinks commented on lib/components/Search.dart:
```diff
@@ -3,16 +3,25 @@ import 'package:veggie_seasons_v2/styles/styles.dart';
 
   @override
   Widget build(BuildContext context) {
+    final OutlineInputBorder border = OutlineInputBorder(
```
Consolidated since it was being duped, more concerned about possible bug mentioned below though.

khanhnwin replied:
Can this be reused in the Styles.dart file? Admittedly I still confuse OutlineInputBorder and BoxDecoration. It _feels_ like they should be the same thing, but I know it's not.

Piinks replied:
Oh for sure. You can put whatever you like in the style file. I noticed there were some box decorations there, but then box decorations elsewhere in the app. I did not go so far as to refactor everything to put consistent box decorations in the style file for reuse throughout the app. That could be a nice cleanup though!

Piinks commented on lib/components/TriviaComponent.dart:
```diff
@@ -41,13 +41,7 @@ class _TriviaComponentState extends State<TriviaComponent> {
       builder: (BuildContext context) {
         return AlertDialog(
           title: const Text('🎉 Good Job!'),
-          content: const SingleChildScrollView(
-            child: ListBody(
-              children: <Widget>[
-                Text('You got the correct answer.'),
-              ],
-            ),
-          ),
```
None of this was necessary

khanhnwin replied:
I don't remember why I even added that...

Piinks commented on lib/components/TriviaComponent.dart:
```diff
@@ -66,32 +60,32 @@ class _TriviaComponentState extends State<TriviaComponent> {
   Widget build(BuildContext context) {
```
I actually crashed when I got to this page, so I tweaked.
Reduced a bit re: padding and sized boxes

khanhnwin replied:
Do you know what specifically caused you to crash? I see the code stylistically reformats some things but not anything that I'd expect to cause a crash!

Piinks replied:
I think it was a render overflow error. I was running on desktop, and after we caught up I learned this was written with a mobile focus.

Piinks commented on lib/components/cards/CardHeader.dart:
```diff
@@ -13,7 +13,7 @@ class CardHeader extends StatelessWidget {
       mainAxisAlignment: MainAxisAlignment.spaceBetween,
       children: [
-        Flexible(
+        Expanded(
```
Expanded is a Flexible subclass. Since this was the only Flexible, and Expanded is used elsewhere, I swapped it out for consistency.

Piinks commented on lib/components/cards/SmallVeggieCard.dart:
```diff
@@ -23,29 +23,27 @@ class SmallVeggieCard extends StatelessWidget {
-            ],
-          ),
+        decoration: $styles.containerStyles.rounded(),
+        padding: EdgeInsets.all($styles.padding.m),
```
Container.padding

Piinks commented on lib/components/cards/SmallVeggieCard.dart:
```diff
@@ -23,29 +23,27 @@ class SmallVeggieCard extends StatelessWidget {
-                  crossAxisAlignment: CrossAxisAlignment.start,
-                  children: [
-                    CardHeader(veggie: veggie, size: size),
-                    Gap($styles.padding.s),
```
Spacing is applied throughout the app with Paddings, SizedBoxes, Containers, and here Gap. I tried to simplify to reduce all the options.

Piinks commented on lib/components/cards/VeggieCard.dart:
```diff
@@ -22,25 +22,21 @@ class VeggieCard extends StatelessWidget {
-            ],
-          ),
+        decoration: $styles.containerStyles.rounded(),
+        padding: EdgeInsets.all($styles.padding.m),
```
Container.padding

Piinks commented on lib/components/cards/VeggieCard.dart:
```diff
@@ -22,25 +22,21 @@ class VeggieCard extends StatelessWidget {
-        decoration: $styles.containerStyles.rounded,
-        child: Padding(
-          padding: EdgeInsets.all($styles.padding.m),
-          child: Column(
-            children: [
-              VeggieImage(veggie: veggie, size: size),
-              Column(
-                crossAxisAlignment: CrossAxisAlignment.start,
```
Column inside of Column not needed with some alignment instead. Also reduce all the Padding widgets

Piinks commented on lib/components/cards/VeggieCardSeasons.dart:
```diff
@@ -15,9 +15,9 @@ class VeggieCardSeasons extends StatelessWidget {
-            padding: size == Size.large
-                ? EdgeInsets.only(left: $styles.padding.s)
-                : EdgeInsets.only(left: $styles.padding.xs),
+            padding: EdgeInsets.only(
```
Tiny simiplification.

khanhnwin replied:
I can't believe I hadn't even considered this as possible!

Piinks commented on lib/components/cards/VeggieImage.dart:
```diff
@@ -20,13 +20,14 @@ class VeggieImage extends StatelessWidget {
-        width: width,
-        height: height,
-        decoration: $styles.containerStyles.rounded,
-        child: ClipRRect(
-          borderRadius: $styles.borderStyle.roundedBorderRadius,
-          child: Image(
-              fit: BoxFit.cover, image: AssetImage(veggie.imageAssetPath)),
```
No need for ClipRRect and such. The BoxDecoration already takes an image.

Piinks commented on lib/main.dart:
```diff
@@ -20,7 +20,7 @@ class VeggieSeasonsApp extends StatelessWidget {
       theme: ThemeData(
         scaffoldBackgroundColor: $styles.colors.background,
-        fontFamily: $styles.text.fontFamily,
+        // fontFamily: $styles.text.fontFamily,
```
The app crashed on this so I removed it just so I could focus on the rest.

Piinks commented on lib/screens/Garden.dart:
```diff
@@ -39,33 +39,27 @@ class _GardenState extends State<Garden> {
-                  },
-                  childCount: snapshot.data!.length,
-                ),
+              sliver: SliverList.builder(
```
builder constructor removes some complexity

Piinks commented on lib/screens/SearchPage.dart:
```diff
@@ -35,33 +35,29 @@ class _SearchPageState extends State<SearchPage> {
+      ),
       body: Padding(
         padding: EdgeInsets.symmetric(horizontal: $styles.padding.l),
-        child: Column(
-          children: [
-            Expanded(
-              child: ListView.builder(
-                shrinkWrap: true,
```
Column, Expanded and shrinkwrap aren't necessary. The ListView already expands to fill the available space.

Piinks commented on lib/screens/VeggieDetail.dart:
```diff
@@ -57,34 +57,92 @@ class _VeggieDetailState extends State<VeggieDetail> {
     });
   }
 
+  Widget _regularComponent(Widget servingInfo) {
```
I pulled this out because inside of the build method it was super indented and hard to read

khanhnwin replied:
Can this be an instance where we might want to refactor this widget as a separate widget instead of a method?

Piinks commented on lib/screens/VeggieDetail.dart:
```diff
@@ -57,34 +57,92 @@ class _VeggieDetailState extends State<VeggieDetail> {
-    var servingInfo =
-        Row(mainAxisAlignment: MainAxisAlignment.spaceBetween, children: [
-      Expanded(
-        flex: 1,
```
flex: 1 is the default, no need to specify on all of these.
A lint would have caught this.

khanhnwin replied:
Just so we have as a reference, the lint is `avoid_redundant_argument_values` but that's not included in flutter_lints atm. Kate's opened an issue here: https://github.com/dart-lang/lints/issues/170

Piinks commented on lib/screens/VeggieDetail.dart:
```diff
@@ -57,34 +57,92 @@ class _VeggieDetailState extends State<VeggieDetail> {
+        ),
+        $styles.spacers.xs,
+        Container(
+          padding: EdgeInsets.all($styles.padding.m),
```
Container.padding

Piinks commented on lib/screens/VeggieDetail.dart:
```diff
@@ -218,7 +206,7 @@ class ServingInfo extends StatelessWidget {
     return Column(
       children: [
         NumberedCircle(number: number),
-        SizedBox(height: $styles.padding.xs),
```
I set up some standard spacers in the $styles object to replaced SizedBoxes that were being used for spacing

Piinks commented on lib/styles/styles.dart:
```diff
@@ -8,15 +8,16 @@ class AppStyle {
-  final TextStyle font = GoogleFonts.lexend();
-  final String? fontFamily = GoogleFonts.lexend().fontFamily;
+  final TextStyle font = const TextStyle();
+  // final String? fontFamily = GoogleFonts.lexend().fontFamily;
```
This was throwing, so I took it out to work on the rest.

khanhnwin replied:
Google Fonts probably required adding allowing network access on macOS! I just opened a PR to update the setup instructions in the readme. https://github.com/leighajarett/veggie_seasons_v2/pull/4

Piinks commented on lib/styles/styles.dart:
```diff
@@ -59,33 +60,41 @@ class _Padding {
 }
 
 class _ContainerStyles {
+  // Why are these defined here, but others defined ad hoc in app?
```
I like having some standard BoxDecorations defined for the app. But it looked like this wasn't consistently used. I'd recommend auditing all the BoxDecorations to see if it can be consolidated further.

Piinks commented on macos/Podfile.lock:
```diff
@@ -3,20 +3,27 @@ PODS:
   - path_provider_foundation (0.0.1):
```
Ignore stuff below here.

Piinks committed:
Fix navigation bar

Piinks committed:
Nit

Piinks committed:
Tab controller cleanup

Piinks commented:
Ah I figured I fiddled too far with it. I can take a look today and test across devices.

Piinks committed:
UPdates

Piinks commented:
Fixed @khanhnwin! 

I also cleaned up the import statements. In my experience, local imports are used for files within the project. You could alternatively create export files for different parts like style and components, but I would probably only recommend that for large enterprise applications. I can push a separate branch that demonstrates this if you like, let me know.

I also replaced `import:material` where I could and used the widgets library instead. Could be interesting to examine the remaining files that are importing material to see if we can identify other areas where basic widgets or functions are not available outside of material.

khanhnwin committed:
Put back GoogleFonts

khanhnwin commented:
Thanks @Piinks! Looks good on my end now :D

> In my experience, local imports are used for files within the project. You could alternatively create export files for different parts like style and components, but I would probably only recommend that for large enterprise applications. I can push a separate branch that demonstrates this if you like, let me know.

"local imports are used for files within the project" This will make for a good "best practice" recommendation! As for the export files, are these similar to barrel files? Why only enterprise applications?

khanhnwin commented on lib/components/cards/VeggieImage.dart:
```diff
@@ -20,13 +21,14 @@ class VeggieImage extends StatelessWidget {
+      width: width,
+      height: height,
+      decoration: $styles.containerStyles.rounded(
+        image: DecorationImage(
```
IIRC - I think the ClipRRect was an artifact of the Image corners not respecting the Container decoration's rounded corner or something like that. But now knowing the decoration can take an image is a big plus!

khanhnwin commented on lib/components/Header.dart:
```diff
@@ -17,38 +18,34 @@ class CustomSliverHeaderDelegate extends SliverPersistentHeaderDelegate {
+            mainAxisAlignment: MainAxisAlignment.end,
+            crossAxisAlignment: CrossAxisAlignment.start,
+            children: [
+              if (progress < 1)
```
Making running notes of your review. Am I interpreting this correctly? "Ternary when there is a true and false case. If for when there’s just a “show THIS widget when X is true…”"

Piinks replied:
I think this one was less about ternary versus `if`.

The other end of the ternary operator in this original code was to insert a SizedBox.shrink(), which is basically nothing. Inline in a list `[]`, the ternary operator can be hard to read, adds to indentation etc.
Since the options were add this AnimatedOpacity - or - add a nothing widget, just removing the unnecessary widget and making it an if for the first case felt easier to understand.

khanhnwin replied:
Thanks for clarifying! Makes sense!

khanhnwin commented on lib/components/toggle.dart:
```diff
@@ -36,25 +34,19 @@ class _ToggleState extends State<Toggle> {
-                    'Trivia',
-                    style: $styles.text.heading3,
-                  ),
+                  label: Text('Trivia', style: $styles.text.heading3),
```
Is there a stylistic reason for not having the trailing comma here and in line 45? I like always having the trailing comma because it makes it readable.

Piinks replied:
Yeah, this is basically just a stylistic preference. If the content fits on one line within a reasonable extent, I prefer to have it be one line rather than spreading this and its short parameters over 4 lines. 
I would have done the same with BorderSide above on l#36, but when I subbed in the `$styles` values it was much longer, so I split it.
I know I have heard all of the indentation and closing parentheses can be fatiguing to readability.

khanhnwin replied:
Got it! That makes sense. I think Bob had mentioned the potential of having dart fix always adding that trailing comma at some point.

Piinks commented:
> "local imports are used for files within the project" This will make for a good "best practice" recommendation!

💯 
Yeah, makes it clear where things are coming from too, is the call coming from inside the house or from a package? 😆 

> As for the export files, are these similar to barrel files? Why only enterprise applications?

I am not sure. So this would look like... let's look at the components directory. Instead of importing every individual file on an ad hoc basis, you could have a `components.dart` file that exports all of the other files in `components/`. Then in practice, you would only have to import the components.dart file.

I mentioned enterprise since they are more likely to have bigger apps with LOTS of files, components and  assets. So exporting subsets of the app would be nice and clean IMO. Since veggie seasons is so small, it doesn't seem necessary.

khanhnwin commented on lib/screens/VeggieDetail.dart:
```diff
@@ -111,91 +169,22 @@ class _VeggieDetailState extends State<VeggieDetail> {
-                  size: Size.large,
-                  veggie: widget.veggie,
-                )
+                VeggieCardSeasons(size: Size.large, veggie: widget.veggie),
```
trailing comma confuses me again ha

khanhnwin reviewed (approved):
Thanks @Piinks! I got a good doc of notes from going through your PR too!

khanhnwin commented:
> makes it clear where things are coming from too, is the call coming from inside the house or from a package? 😆

Unfortunately, it's usually coming from inside the house for me 😆

> I am not sure. So this would look like... let's look at the components directory. Instead of importing every individual file on an ad hoc basis, you could have a components.dart file that exports all of the other files in components/. Then in practice, you would only have to import the components.dart file.

> I mentioned enterprise since they are more likely to have bigger apps with LOTS of files, components and assets. So exporting subsets of the app would be nice and clean IMO. Since veggie seasons is so small, it doesn't seem necessary.

Gotcha! That's what I know as barrel files, we're on the same page there! I actually formatted my files that exact way because I was lazy with importing individual files. It's a bit extra work for little return on a small app like Veggie Seasons, but there's no reason _to not_ use that components.dart otherwise, right?

khanhnwin commented:
Friendly ping for review @leighajarett!
